### ISSUE TITLE
what is the override config for https://github.com/weluse/hyperagent ?

### ISSUE THREAD

**Issue by docteurklein** [jspm/registry#483]

**Comment by Munter** [comment]
It's probably going to be hard to deal with that `Hyperagent.configure(...)` thing. But it should be possible to override in a way so the assumed defaults `window.$` and `window.Q` are available. I'll take a look at it

**Comment by docteurklein** [comment]
Thanks a lot for your response :)

I guess I can live with a manual invokation of configure if necessary: In fact configure is just optional.

My biggest concern was how to configurethe shims and co.
I tried hard to modify my jspm config inside packages.json, but it looked like the generated config.js didn't care at all about my overrides.
Anyway, thanks :)

**Comment by Munter** [comment]
Took me a while to figure out how overrides even affect things. I think this need a more explicit workflow an cause/effect explanation. Overrides actually mostly add comment annotations in the downloaded source code, which can then instruct the loader to do the right thing. This is cool because it doesn't clutter up your local config, but on the other hand without a clear explanation of what it does it feels a bit like black magic.

**Comment by Munter** [comment]
So this gives me the correct dependencies in that it also installs jquery and q when installing, it loads jquery and q when loading the package. But I can't figure out the missing config to force jquery and q to expose globals, which will then be picked up by hyperagent. It picks up jquery because jquery apparently always leaks itself to `$` in window (bad jquery), but Q doesn't do that.

```
jspm install hyperagent=github:weluse/hyperagent -o "{ main: 'dist/hyperagent.js', format: 'global', dependencies: { jquery: '*', q: 'npm:q' }, shim: { 'dist/hyperagent': { deps: ['jquery', 'q'] } } }"
```

**Comment by docteurklein** [comment]
wow you rock!

Maybe we should override Q, with a shim and a global format ?

**Comment by docteurklein** [comment]
BTW @Munter I don't know if it helps, but both BAMD and cjs are present in the dist folder.

**Comment by Munter** [comment]
That would be bad practice imo. Configuring another package to default to doing something different. It would probably be best if the needed configuration was encapsulated in hyperagent itself. The behavior we are looking for is something along these lines:

```js
System.import('jquery')
  .then($ => { window.$ = $})
  .then(() => { return System.import('q') })
  .then(q => { window.Q = q })
  .then(() => [ return System.import('hyperagent') })
  .then(hyperagent => {
    // Finally!
  });
```

**Comment by Munter** [comment]
Yeah, I saw the other formats are available. But since hyperagent uses what I would consider an antipattern to define dependencies, or even optional dependencies, they won't work. It's still picking `$` and `Q` from the window scope for the defaults. And the only way I know how to instruct SystemJS to leak globals from other libraries is to force hyperagent to be global.

@guybedford Any inputs here?

**Comment by Munter** [comment]
Allright this is working for me.

jspm install override:
```json
{
  "main": "dist/hyperagent.js",
  "format": "global",
  "dependencies": {
    "jquery": "*",
    "q": "npm:q"
  },
  "shim": {
    "dist/hyperagent": {
      "deps": [
        "jquery",
        "q"
      ]
    }
  }
}
```

config.js:
```js

System.config({
  "meta": {
    "github:weluse/hyperagent@*": {
      "globals": {
        "$": "jquery",
        "Q": "q"
      }
    }
  }
});
```

So the question is. How can I make the override have a similar effect as what I am setting in config.js?

**Comment by docteurklein** [comment]
That is super cool! Thanks a lot. And I'm curious of the awswer too. Apparently overrides don't modify anything (but map?) in config.js.

**Comment by docteurklein** [comment]
Also I still have an issue with URI and URITemplate.

**Comment by Munter** [comment]
Right I missed those. [this is the correct override](https://github.com/Munter/registry/blob/hyperagent/package-overrides/github/weluse/hyperagent%400.1.0.json)

And this is the correct meta config:
```js
System.config({
  "meta": {
    "github:weluse/hyperagent@*": {
      "globals": {
        "$": "jquery",
        "Q": "q",
        "URI": "uri",
        "URITemplate": "uri/src/URITemplate"
      }
    }
  }
});
```

Has the maintainer of this package ever heard of package managers or module loaders? This module could sure use an overhaul

**Comment by docteurklein** [comment]
Thanks a lot for your effort!

Now, should I install the package from jspm and the override will be automatic ?

    jspm install hyperagent=jspm:hyperagent

**Comment by Munter** [comment]
just `jspm install hyperagent` will do. Unless you specifically want to map it to a different name locally

**Comment by docteurklein** [comment]
but the meta config should still be put manually, right?

**Comment by Munter** [comment]
Correct. Overrides can' do that. Yet. I'll see if I can take a look at enabling that in a PR in the near future
